### ISSUE TITLE
stats: tag specifier not working

### ISSUE THREAD

**Issue by moderation** [envoyproxy/envoy#2538]
*Title*: tag specifier not working

*Description*: From Slack...

> I have two Envoys running on a single node. One is a front proxy and one is a sidecar service proxy. The front proxy listener is TLS. Both are configured to send statsd metrics to a statsd process on the host. The issue is that `stat_prefix` doesn't apply to `envoy_server_days_until_first_cert_expiring` so my Grafana dashboard occasionally shows `2147483647` (no certificate on the sidecar Envoy) and occasionally shows `54` which is the correct value on the TLS listener Envoy. Is there a way to work around this and be able to pull a specific value from a statsd service?

- https://envoyproxy.slack.com/archives/C78M4KW76/p1517785258000052?thread_ts=1517695444.000074&cid=C78M4KW76
- https://envoyproxy.slack.com/messages/C78M4KW76/convo/C78M4KW76-1517855099.000259/

*Relevant Links*:
- https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/metrics/v2/stats.proto#config-metrics-v2-tagspecifier
- https://github.com/envoyproxy/envoy/issues/1975
- https://github.com/envoyproxy/envoy/pull/2264
- https://github.com/envoyproxy/envoy/pull/2357

/cc @taiki45 @ccaraman 

*Repro steps*:
Add the following to config

```
  "stats_config": {
    "stats_tags": [
      {
        "tag_name": "server.missing_cert",
        "fixed_value": "server.days_until_first_cert_expiring"
      }
    ]
  },
```
*Expected result*:
Expect to see an entry of `server.missing_cert` in `/stats` output (or in statsd output).

*Observed result*: no change to `/stats` output.

**Comment by taiki45** [comment]
`envoy.statsd` sink does not support tagged metrics. `envoy.dog_statsd` sink supports tagged metrics with DogStatsD extension: https://docs.datadoghq.com/developers/dogstatsd/. I didn't try but `envoy.metrics_service` may also support that.

**Comment by taiki45** [comment]
To avoid the confusing, is it better to describe tagged metrics support in document? I'm thinking which place will be better. Maybe at https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/metrics/v2/stats.proto#config-metrics-v2-statssink?

**Comment by mattklein123** [comment]
@taiki45 it would be really great IMO to actually discuss tags in the arch overview: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/statistics. Perhaps you could work with @mrice32 on this since he has done a lot of work in this area? (He can help w/ English copy editing also if needed).

**Comment by taiki45** [comment]
Yeah, I also think the statics arch overview document needs update. I'll be working on it. Once the draft is ready, it's great to be reviewed with @mrice32 and others.

**Comment by mattklein123** [comment]
Thank you @taiki45!

**Comment by ccaraman** [comment]
Can the config loading code path be updated as well? If statsd is used and tags are specified, config loading should fail.

**Comment by taiki45** [comment]
+1 to adding config validation logic around stats to avoid confusing, but as @mrice32 described we need a discussion about some topics around stats config. I'll file a new issue to discuss and track them.

**Comment by ofek** [comment]
Is there any doc currently that explains tag extraction?

**Comment by mrice32** [comment]
The docs have a section on how it works: https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/metrics/v2/stats.proto.  Let us know if there's anything unclear after reading that, and we can try to improve it.

**Comment by ofek** [comment]
@mrice32 Thanks! What would the equivalent yaml section be?

**Comment by mrice32** [comment]
I'm not sure I understand the question.  Others can correct me if I'm wrong, but I believe that all of our configuration examples (both v1 and v2) in the docs are JSON rather than YAML.  So I don't think there exists an equivalent YAML section in the docs.

**Comment by ofek** [comment]
I guess what I'm asking how to use that in my current config, which is yaml.

**Comment by mrice32** [comment]
Here's the json: 
```
{
  ...
  "stats_config": {
  	"stats_tags": [
      {
  			"tag_name": "my_custom_tag",
  			"regex": "^someregex"
			}
    ],
  	"use_all_default_tags": true
  }
  ...
}
```

and the YAML:

```
---
...
stats_config:
  stats_tags:
  - tag_name: my_custom_tag
    regex: "^someregex"
  use_all_default_tags: true
...
```
